File: Goweb/docs/src/utils/feedback.ts
import { layer } from "@layui/layui-vue";
import { useAppStore } from "../store/app";

// ============================================================================
// Feedback utility — wrap layer API theo quy tắc Layer.md
// Dùng chung cho toàn app, tránh config rải rác.
// ============================================================================

/** Shade style theo theme: đen 70% (light) / trắng 30% (dark) */
function getShadeStyle(): Record<string, string> {
  try {
    const appStore = useAppStore();
    if (appStore.theme === "dark") {
      return { backgroundColor: "rgba(255, 255, 255, 0.3)" };
    }
  } catch {
    // store chưa sẵn sàng
  }
  return { backgroundColor: "rgba(0, 0, 0, 0.7)" };
}

// ── Toast messages ──────────────────────────────────────────────────────────

/** Success toast — 3s tự biến mất (Layer.md: 3-4 giây) */
export function msgSuccess(text: string, time = 3000) {
  return layer.msg(text, { icon: 1, time });
}

/** Error toast — 4s tự biến mất (Layer.md: 4-5 giây) */
export function msgError(text: string, time = 4000) {
  return layer.msg(text, { icon: 2, time });
}

/** Info toast — 3s */
export function msgInfo(text: string, time = 3000) {
  return layer.msg(text, { icon: 0, time });
}

/** Warning toast — 5s (Layer.md: 5-7 giây) */
export function msgWarn(text: string, time = 5000) {
  return layer.msg(text, { icon: 7, time });
}

// ── Loading ─────────────────────────────────────────────────────────────────

/** Hiện loading spinner, trả về ID để close sau */
export function showLoading() {
  return layer.load(0, {
    shadeStyle: getShadeStyle(),
    shadeOpacity: "1",
  });
}

/** Đóng loading */
export function closeLoading(id: string | number) {
  layer.close(id as any);
}

// ── Confirm dialog ──────────────────────────────────────────────────────────

interface ConfirmOptions {
  yesText?: string;
  noText?: string;
}

/** Confirm dialog — trả Promise<boolean> */
export function confirm(
  message: string,
  opts: ConfirmOptions = {}
): Promise<boolean> {
  const { yesText = "Xác nhận", noText = "Huỷ" } = opts;
  return new Promise((resolve) => {
    layer.confirm(message, {
      shadeStyle: getShadeStyle(),
      shadeOpacity: "1",
      btn: [
        {
          text: yesText,
          callback: (id: any) => {
            layer.close(id);
            resolve(true);
          },
        },
        {
          text: noText,
          callback: (id: any) => {
            layer.close(id);
            resolve(false);
          },
        },
      ],
    });
  });
}

// ── Wrap async operation (loading + success/error) ──────────────────────────

interface RunOptions {
  loadingText?: string;
  successMsg?: string;
  errorMsg?: string | ((err: any) => string);
}

/**
 * Wrap một async operation với loading + feedback tự động.
 *
 * @example
 * const result = await feedback.run(
 *   () => api.post("/api/agents", data),
 *   { successMsg: "Thêm thành công", errorMsg: "Thêm thất bại" }
 * );
 */
export async function run<T>(
  fn: () => Promise<T>,
  opts: RunOptions = {}
): Promise<T | null> {
  const loading = showLoading();
  try {
    const result = await fn();
    if (opts.successMsg) {
      msgSuccess(opts.successMsg);
    }
    return result;
  } catch (err: any) {
    const msg =
      typeof opts.errorMsg === "function"
        ? opts.errorMsg(err)
        : opts.errorMsg ||
          err.response?.data?.message ||
          err.response?.data?.error ||
          "Thao tác thất bại";
    msgError(msg);
    return null;
  } finally {
    closeLoading(loading);
  }
}

// ── Confirm rồi chạy (delete, logout, ...) ─────────────────────────────────

interface ConfirmRunOptions extends RunOptions {
  confirmMsg: string;
  yesText?: string;
  noText?: string;
}

/**
 * Confirm trước, nếu OK thì chạy async operation với loading + feedback.
 *
 * @example
 * await feedback.confirmRun({
 *   confirmMsg: 'Xác nhận xoá đại lý "ABC"?',
 *   yesText: "Xoá",
 *   successMsg: "Đã xoá",
 * }, () => api.delete(`/api/agents/${id}?mode=destroy`));
 */
export async function confirmRun<T>(
  opts: ConfirmRunOptions,
  fn: () => Promise<T>
): Promise<T | null> {
  const confirmed = await confirm(opts.confirmMsg, {
    yesText: opts.yesText,
    noText: opts.noText,
  });
  if (!confirmed) return null;
  return run(fn, opts);
}

/** Shade style getter — dùng cho declarative lay-layer components */
export { getShadeStyle };

const feedback = {
  msgSuccess,
  msgError,
  msgInfo,
  msgWarn,
  showLoading,
  closeLoading,
  confirm,
  run,
  confirmRun,
  getShadeStyle,
};

export default feedback;

File: GoDesktop/src/utils/feedback.cpp
#include "utils/feedback.h"
#include "core/toast_widget.h"
#include "core/confirm_dialog.h"
#include <stdexcept>

Feedback::Feedback(QWidget* parent, ThemeManager* theme)
    : m_parent(parent)
    , m_theme(theme)
{
}

// ── Toast messages ──────────────────────────────────────────────────────────

void Feedback::msg_success(const QString& text, int duration_ms)
{
    ToastWidget::show_toast(m_parent, text, ToastWidget::Success, duration_ms);
}

void Feedback::msg_error(const QString& text, int duration_ms)
{
    ToastWidget::show_toast(m_parent, text, ToastWidget::Error, duration_ms);
}

void Feedback::msg_warn(const QString& text, int duration_ms)
{
    ToastWidget::show_toast(m_parent, text, ToastWidget::Warning, duration_ms);
}

void Feedback::msg_info(const QString& text, int duration_ms)
{
    ToastWidget::show_toast(m_parent, text, ToastWidget::Info, duration_ms);
}

// ── Loading ─────────────────────────────────────────────────────────────────

ToastWidget* Feedback::show_loading(const QString& text)
{
    const QString msg = text.isEmpty() ? QStringLiteral("Đang xử lý...") : text;
    return ToastWidget::show_toast(m_parent, msg, ToastWidget::Loading, 0);
}

void Feedback::close_loading(ToastWidget* loading)
{
    if (loading) {
        loading->close_toast();
    }
}

// ── Confirm dialog ──────────────────────────────────────────────────────────

bool Feedback::confirm(const QString& message,
                       const QString& yes_text,
                       const QString& no_text)
{
    return ConfirmDialog::ask(m_parent, m_theme, message, yes_text, no_text);
}

// ── Wrap async pattern ──────────────────────────────────────────────────────

void Feedback::run(std::function<void(ToastWidget*)> fn,
                   const QString& success_msg,
                   const QString& error_msg)
{
    auto* loading = show_loading();
    try {
        fn(loading);
        close_loading(loading);
        if (!success_msg.isEmpty()) {
            msg_success(success_msg);
        }
    } catch (const std::exception& e) {
        close_loading(loading);
        const QString msg = error_msg.isEmpty()
            ? QString::fromUtf8(e.what())
            : error_msg;
        msg_error(msg);
    } catch (...) {
        close_loading(loading);
        msg_error(error_msg.isEmpty()
            ? QStringLiteral("Thao tác thất bại")
            : error_msg);
    }
}

bool Feedback::confirm_run(const QString& confirm_msg,
                           std::function<void(ToastWidget*)> fn,
                           const QString& success_msg,
                           const QString& error_msg,
                           const QString& yes_text,
                           const QString& no_text)
{
    if (!confirm(confirm_msg, yes_text, no_text)) {
        return false;
    }

    auto* loading = show_loading();
    try {
        fn(loading);
        close_loading(loading);
        if (!success_msg.isEmpty()) {
            msg_success(success_msg);
        }
        return true;
    } catch (const std::exception& e) {
        close_loading(loading);
        const QString msg = error_msg.isEmpty()
            ? QString::fromUtf8(e.what())
            : error_msg;
        msg_error(msg);
        return false;
    } catch (...) {
        close_loading(loading);
        msg_error(error_msg.isEmpty()
            ? QStringLiteral("Thao tác thất bại")
            : error_msg);
        return false;
    }
}
